articles/designing_saas.rst

```diff
@@ -0,0 +1,188 @@
+Designing your SaaS Database for Scale with Postgres
+####################################################
+
+If you’re building a SaaS application, you probably already have the
+notion of tenancy built in your data model. Typically, most information
+relates to tenants / customers / accounts and your database tables
+capture this natural relation.
+
+With smaller amounts of data (10s of GB), it’s easy to throw more
+hardware at the problem and scale up your database. As these tables grow
+however, you need to think about ways to scale your multi-tenant
+database across dozens or hundreds of machines.
+
+After our blog post on `sharding a multi-tenant app with
+Postgres <https://citusdata.com/blog/2016/08/10/sharding-for-a-multi-tenant-app-with-postgres/>`__,
+we received a number of questions on architectural patterns for
+multi-tenant databases and when to use which. At a high level,
+developers have three options:
+
+1. Create one database per tenant
+2. Create one schema per tenant
+3. Have all tenants share the same table(s)
+
+The option you pick has implications on scalability, how you handle data
+that varies across tenants, isolation, and ease-of-maintenance. And
+these implications have been discussed in detail across many
+`StackOverflow
+questions <http://stackoverflow.com/search?tab=votes&q=multi-tenant%20database>`__
+and `database
+articles <https://msdn.microsoft.com/en-us/library/aa479086.aspx>`__.
+So, what is the best solution?
+
+In practice, each of the three design options -with enough effort- can
+address questions around scale, data that varies across tenants, and
+isolation. The decision depends on the *primary* dimension you’re
+building/optimizing for. The tldr:
+
+-  If you’re building for scale: Have all tenants share the same
+   table(s)
+-  If you’re building for isolation: Create one database per tenant
+
+In this article, we’ll focus on the scaling dimension, as we found
+that more users who talked to us had questions in that area. (We also
+intend to describe considerations around isolation in a follow-up blog
+post.)
+
+To expand on this further, if you’re planning to have 5 or 50 tenants in
+your B2B application, and your database is running into scalability
+issues, then you can create and maintain a separate database for each
+tenant. If however you plan to have thousands of tenants, then sharding
+your tables on a tenant\_id/account\_id column will help you scale in a
+much better way.
+
+.. figure:: ../images/articles-saas-fig-1.png
+   :alt: Sharding tables for multi-tenant applications
+
+   Sharding tables for multi-tenant applications
+
+Common benefits of having all tenants share the same database are:
+
+Resource pooling (reduced cost): If you create a separate database for
```

Comment:
We have a few words / sentences that are underlined in the original blog post. (I think "resource pooling", "if you shard ..." sentence, "ease of maintenance" and maybe 1-2 others).

Could we underline these phrases or sentences?